Add missing docstrings for `EquivalenceLibrary`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits attempt to fix #13258 by adding all of the missing docstrings that were lost after #12585.

### Details and comments
Add missing docstrings in the following methods:
- Python (PyO3):
    - `EquivalenceLibrary` struct class.
    - `EquivalenceLibrary::get_graph` method.
    - `EquivalenceLibrary::py_keys` method.
    - `EquivalenceLibrary::py_node_index` method.
- Rust:
    - `EquivalenceLibrary::keys` method.

raynelfss committed:
Initial: Add missing docstrings for `EquivalenceLibrary`

kevinhartman commented on crates/accelerate/src/equivalence.rs:
```diff
@@ -340,6 +340,8 @@ impl ToPyObject for CircuitFromPython {
 type GraphType = StableDiGraph<NodeData, Option<EdgeData>>;
 type KTIType = IndexMap<Key, NodeIndex, RandomState>;
 
+/// A library providing a one-way mapping of Gates to their equivalent
```
I think we should either reference the classes with code refs or just call them "gates" and "circuits" :)

raynelfss replied:
I'd say to go with "gates" and "circuits" because `PyO3` will reuse this docstring with Python, which may not be displayed properly.

kevinhartman replied:
I think we're in agreement here, but I meant that it'd be better to not use symbol-sounding names like "QuantumCircuits" and "Gates" with a capital "G". Just call them "gates" and "circuits", verbatim.

kevinhartman replied:
Also, since this is a Python class, you can use Sphinx code references if you want, instead. Like 

```
:class:`.QuantumCircuit`
```

They should turn into hyperlinks in the Python docs.

mtreinish replied:
I agree with Kevin here it should be `gates` or we do a class ref if it's a capital `G`. Although I realize the original text had `Gates` and `QuantumCircuits` without a link. We can do better here.

```suggestion
/// A library providing a one-way mapping of gates to their equivalent
```

kevinhartman commented on crates/accelerate/src/equivalence.rs:
```diff
@@ -664,6 +687,7 @@ impl EquivalenceLibrary {
         self.key_to_node_index.contains_key(&key)
     }
 
+    /// Returns an iterator with all the Keys in the `EquivalenceLibrary`.
```
For Rust doc strings, we should use the Rust-style for code refs, e.g. `[EquivalenceLibrary]`.

raynelfss committed:
Fix: Use code-refs to refer to structs in rust.

kevinhartman commented on crates/accelerate/src/equivalence.rs:
```diff
@@ -682,13 +704,13 @@ impl EquivalenceLibrary {
     ///
-    /// # Arguments:
-    /// * `key`: The `Key` to look for.
+    /// ## Arguments:
```
Is there a reason to go with H2 here?

raynelfss replied:
Nothing more than just me thinking H1 looks a bit too big here.

kevinhartman replied:
Heh, understandable. Regardless, it looks like the Rust standard library uses H1 for a first level section, but it also looks like documenting arguments separately like we do in Python is actually not hip in Rust. 

I would recommend that we just remove the `Arguments` and `Returns` sections on all of these methods altogether. To describe arguments that aren't self-explanatory, add a second paragraph after the summary sentence to describe the usage.

Here's what the Rust standard library uses as their convention: https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md#appendix-a-full-conventions-text

kevinhartman reviewed (commented):
The Python docs look good. I left one comment recommending we use lowercase generic terms in cases where we don't want to make a code reference to a type. In general, prefer `circuit` or `` :class:`.QuantumCircuit` `` over `QuantumCircuit`. This is a fairly standard practice, e.g. [Rust gives this guidance in its standard library style guide](https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md#referring-to-types). The rationale that comes to mind is that if we know the exact type in question, we might as well use a code reference.

As for the Rust docs, I'd recommend we drop all of the `Args` and `Returns` sections, and stick with just a summary sentence, followed by an optional paragraph that describes any usage details. It's just not standard practice in Rust to list out arguments like we see in Python docs. Instead, the Rust community favors a cohesive usage description of the function as a whole when arguments aren't obvious, along with an `Examples` section, ideally for every method to provide usage examples. It's probably alright to skip the `Examples` section for most of the methods you've got in this PR, since the usage is self-explanatory and this is not yet a public interface.

raynelfss committed:
Remove `Arguments` and `Returns` sections.
- Add docs links based on @kevinhartman's comments.
- Add sphinx friendly labels in docstrings.

mtreinish commented on crates/accelerate/src/equivalence.rs:
```diff
@@ -436,8 +440,8 @@ impl EquivalenceLibrary {
-    /// Gets the set of QuantumCircuits circuits from the library which
-    /// equivalently implement the given Gate.
+    /// Gets the set of :class:`.QuantumCircuit` instances circuits from the
+    /// library which equivalently implement the given Gate.
```
```suggestion
    /// library which equivalently implement the given :class:`.Gate`.
```

mtreinish committed:
Apply suggestions from code review

mtreinish reviewed (approved):
LGTM, thanks for fixing this.
